### PR TITLE
fabric: Define CQ entry fields

### PR DESCRIPTION
**Pull request by shefty**
Formally define the meaning of the CQ entry fields, including
when the fields are valid.

Signed-off-by: Sean Hefty <sean.hefty@intel.com>

### CODE REVIEW

**Comment by shefty** [comment]
@bturrubiates @sayantansur @jithinjosepkl @a-ilango @j-xiong 

Please review to ensure that the documentation makes sense and that the providers are actually following what is documented.  If any provider is not (or cannot), please let me know, so that we can document the fields correctly.

**Comment by j-xiong** [comment]
This documentation looks good to me and matches what the psm provider is doing now. The psm provider currently sets some extra fields for certain operations but I think that's fine since those fields are considered not defined for those operations. Work is in progress to remove those unnecessary field settings to save a cycle  or two. 

There is a typo in the last line --  "untils".

**Comment by shefty** [comment]
Thanks - typo fixed

File: man/fi_cq.3.md [hunk 1]
```diff
@@ -382,6 +383,72 @@ or fi_cq_sreadfrom.  This may be used to wake-up a thread
...
+  _Completion Flags_ section below lists valid flag values.  Flags are
+  set for all relevant completions.
+
+*len*
+: This len field only applies to completed receive operations.  It indicates
+  the size of received message data -- i.e. the number of bytes of data that
+  was actually placed into the associated receive buffer.  If an endpoint has
+  been configured with the FI_MSG_PREFIX mode, the len also reflects the size
+  of the prefix buffer.
+
```
Comment (bturrubiates) [inline man/fi_cq.3.md]: I'm not sure how the `usnic` provider handles this, looking into it.

The `fi_send` and `fi_recv` calls don't take a size that includes the prefix buffer, so should completion fields reflect that behavior?

Comment (shefty) [inline man/fi_cq.3.md]: > The fi_send and fi_recv calls don't take a size that includes the prefix
> buffer, so should completion fields reflect that behavior?

I didn't define the len for sends, only for receives.  I wasn't sure what usnic did, so I just defined something, figuring I had a 50/50 chance of getting it correct.  :)  I agree that the completion should match the send/recv calls.

Are you sure that send/recv don't take the prefix size?  That seems unfortunate, as it makes it impossible to use the prefix with an iov.

**Comment by jithinjosepkl** [comment]
It looks good to me also. Afaik, sockets provider follows this documentation.

File: man/fi_cq.3.md (hunk 1 above)
Comment (bturrubiates) [inline man/fi_cq.3.md]: Actually, I might be mistaken. I’ll look into it.

- Ben

**Comment by a-ilango** [comment]
I think verbs provider supports some of these fields. The ones which aren't supported right now are related to tagged messaging and multi recv: buf, tag. 

I believe verbs provider can't support olen. libibverbs has an error code when there isn't enough receive buffer space but I don't see it reporting how much more space is needed.

**Comment by shefty** [comment]
Being able to report olen is subject to the endpoint not being thrown into an error state when an overflow occurs.  Maybe that field only makes sense for reliable, unconnected endpoints.  ?

**Comment by a-ilango** [comment]
Yes, in case of overflow error or any other error that occurs when reading a CQ, the endpoint would be placed in error state. I checked with verbs spec. Even for reliable, unconnected endpoints, the QP would be moved to error state when there is a receive queue completion error.

**Comment by bturrubiates** [comment]
I agree on the behavior as described in the page. Should `fi_msg.3` also be updated to include that the len given to the prefix operations should include the prefix size?

**Comment by shefty** [comment]
I can update fi_msg with the len.  Is that what usnic is actually expecting?

**Comment by shefty** [comment]
From the fi_getinfo man page, regarding FI_MSG_PREFIX:

  Additionally, applications may receive
  provider generated packets that do not contain application data.
  Such received messages will indicate a transfer size of 0 bytes.

This wording implies that the prefix is not included in the receive completion...

**Comment by bturrubiates** [comment]
I'm not sure, I'm working on fixing some sizing inconsistencies. Whatever behavior we decide on should be well documented.

**Comment by jsquyres** [comment]
To be clear: the usnic provider, when used with `FI_MSG_PREFIX`, requires that both the send and receive lengths passed by the app include the prefix length.  Hence, `fi_ud_pingpong` in fabtests, when it only passes an `fi_send()` / `fi_recv()` length of 16 (when used with `FI_MSG_PREFIX`), is erroneous.

My $0.02: we should be consistent on either always including the prefix length or always excluding the prefix length in send, receive, and completion lengths.

FWIW, the GRH header length is included in UD verbs receives (i.e., you never receive a UD message with length less than 40).  Per my first remarks: the usnic provider is implemented in this spirit.

**Comment by shefty** [comment]
I have another patch.  I don't see a real alternative for send/recv except to require the user to specify the prefix buffer the same as any other buffer.  Otherwise, trying to deal with an IOV where iovec[0] == prefix buffer makes little sense.

I agree that the completion length should include the prefix length as well.  The man page I quoted suggests otherwise, but I'm willing to just reword the page to make it the way we want.  But I do want to know what usnic actually does, so we don't document a breakage without consent.  :)

**Comment by jsquyres** [comment]
@shefty Ok, cool -- I think we're in agreement.  ud_pignpong.c will also need to be updated.

**Comment by goodell** [comment]
FWIW, `olen` could be supported by the usnic hardware, but I think software support currently isn't plumbed through to make this addition trivial for `EP_DGRAM`.  It should be easier for us to add this for our `EP_MSG` and `EP_RDM` support.

When I looked at the psm provider, it looks like it is reporting `0` for `olen` in all cases, except for in `_psmx_tagged_peek` where it sets `olen` to the full `psm_status.msg_length`.  I didn't spend enough time trying to follow the flow of the psm provider code around this area, but it would probably be worth a double check to make sure that usage is consistent with this new man page text.

**Comment by j-xiong** [comment]
The psm  provider does set olen appropriately, please take a look at the function psmx_cq_create_event_from_status(), that's where completions for receives are generated. Notice that non-zero olen only happens for non-zero error code.

**Comment by shefty** [comment]
I updated the PR.  The only change was to fi_getinfo to allow the provider to return a size == msg_prefix_size when no user data is present.

File: man/fi_cq.3.md [hunk 2]
```diff
@@ -382,6 +383,72 @@ or fi_cq_sreadfrom.  This may be used to wake-up a thread
...
+  
+*flags*
+: This specifies flags associated with the completed operation.  The
+  _Completion Flags_ section below lists valid flag values.  Flags are
+  set for all relevant completions.
+
+*len*
+: This len field only applies to completed receive operations.  It indicates
+  the size of received message data -- i.e. the number of bytes of data that
+  was actually placed into the associated receive buffer.  If an endpoint has
```
Comment (goodell) [inline man/fi_cq.3.md]: minor wording suggestion: `i.e., how many data bytes were actually`

**Comment by shefty** [comment]
Updated with Dave's suggestion.  Plus I pushed a second patch that updates fi_msg.3 to indicate that the prefix buffer cannot be split between iovec's, and state that the size includes the prefix buffer length.

**Comment by shefty** [comment]
Opened issue #1142 to track a potential issue with the completion flags.

**Pull request merged**